### ISSUE TITLE
Use /LOG switch in scheduled task by default?

### ISSUE THREAD

**Issue by ghost** [ScriptTiger/Unified-Hosts-AutoUpdate#49]
On several occasions now, I have found that my hosts are not updating, because of the script not updating. Usually, this script "just works" and so I don't really notice it until something is wrong.... and then I go to the scheduled tasks view and get some cryptic MS hex error code.

Then I have to open a command prompt as an admin, change to the correct directory, copypaste the command from three different text fields in the scheduled tasks GUI,  change to the correct directory, paste in the two parts of the command, to run it to see what's happening, to try and troubleshoot it. It's no big deal.... but it is a little cumbersome.

Today I noticed that there is an option to output a log file. I've already added that to my scheduled task manually, but it strikes me that it would be nice to have that as a default. Just makes it a lot easier to see what's going on when the script does its magic... and especially, when it doesn't :)

Edit: Ironically, the failure I had, made me miss [this update](https://github.com/ScriptTiger/Unified-Hosts-AutoUpdate/issues/38) which would have alerted me to the failures XD

**Comment by ScriptTiger** [comment]
The logging and error codes are both relatively new features, so I am glad you like it and thank you for the positive feedback on that. I will keep this issue open for now and get back to it when I have some free time. The issue for me is log management and making sure only a finite amount of data is stored by default, but still be enough data to be useful. Would just having a dump of only the last run be helpful by default?

Right now it allows for user-specified files, so you could make it dump to a different file depending on the day, month, year, or whatever you want, but it's your responsibility to manage those logs after you set it up. So by default it just needs to be a safe option that gives you enough data if you need it, but also doesn't lend itself to building up to infinite if the user isn't monitoring and managing it, which would be an issue for all non-technical users that just set it and forget it.

**Comment by ghost** [comment]
> The logging and error codes are both relatively new features, so I am glad you like it and thank you for the positive feedback on that.

I definitely like it and thank you for adding it :) I didn't realise they were new.

> Would just having a dump of only the last run be helpful by default?

Yeh that would be great. 

Now this is not to say that I have it all figured out but the way I pictured it worked was basically what you said - it logs and overwrites the log each time it runs, unless you specify something custom that provides unique filenames for each run. As you say, that keeps it sane for dummies like me who set it, forget it, and forgot to RTFM (LOL my bad!) but it also leaves room for power users who want to follow it more closely and don't mind managing diskspace/clearing old logs/concatenating multiple log files into a single file/etc.

I'm glad you asked me this question because I just now looked and my log file has already grown to have three entries in it hahaha :) It's funny for me because I can laugh at myself but as a programmer I also know that if I wrote this, someone who didn't RTFM, would blame me, the programmer, when they forgot to read the manual and filled up their C drive!

**TLDR**
So, to put this into the context of the present source code, at the moment we have 
`>>%DEV%`
but I think it might be best to have 
`>%DEV%`
As the first entry, which will overwrite the existing file, and then allow 
`>>%DEV%` to fill it up.
And then if people want to have more than just the previous run in the log file, they can easily manage that by copying/renaming/etc the log file, after this script runs.

I guess, if you wanted to be really fancy, you could include the log size management/concatenation/datestamping/etc, all within this script. Being a linux kind of guy I'm more inclined to the "one tool for each job" approach and it's still very easy to have all that log management in a separate batch script, but it would make this script far more complex. It's up to you what you think is best, I'm just sharing my perspective :)

**Comment by ScriptTiger** [comment]
@xcasxcursex, @ohadschn, I just committed a proposal for the 1.39 release which addresses minimal default logging, which is only a log of the most recent scheduled task (quiet mode)  only. `/log` still works for persistent logging and plays nicely with the new changes, check the new README for more details. I also updated some other things here and there to improve download efficiency and some other things. Please test everything thoroughly and let me know if everything works for you. After you confirm everything, I'll push out the update to everyone else.

And just as a reminder. I know you guys probably already know this because I say it every proposal, but I'll say it anyway. To test a proposed version, download the latest script manually and edit it to change the version number back one, otherwise it will get overwritten by the latest stable release.

I look forward to getting everyone's feedback on this!

**Comment by ghost** [comment]
Haven't done the troubleshooting on this yet but I thought I should post immediately so I might save someone a download - I got errors with this. Be back momentarily with more info.

**Comment by ghost** [comment]
Yeh I had a feeling it was my old thing with the path with spaces.

Wrapping `%LOG%` in quotes to `"%LOG%"` as such: (just these three instances, line 752, 759 and 760)

    rem Function to handle script output
    :Echo
    echo %~1>con
    echo %DATE% @ %TIME%: %~1>>"%LOG%"
    shift
    if not "%~1"=="" goto Echo
    exit /b

    rem Function to log executions
    :Execute
    echo %DATE% @ %TIME%: Executing: %*>>"%LOG%"
    %*>>"%LOG%" || exit /b 1
    exit /b

Got it to behave. I'm not sure if this was the best approach? I did see some comparisons using those vars above, and I don't really know if this is the right way. I'll leave it to the professionals ;)
Anyway, that works either when run as a scheduled task, with or without the /LOG switch, and also as run manually from the prompt.

Edit: By the way, without this, the :Execute routine failed, and I got the "Neither BITS nor PowerShell are installed" message.
Helps if I actually include the error message, right? LOL sorry about that.



Going offtopic here......... Feel free to tune out at this point ;)
I don't want to create even more work for you here but one thing I noticed was that the previous and current/next logs tend to run into each other a little in the log file, eg:

    Mon 16/03/2020 @ 19:06:56.29: Cleaning temporary files...
    Mon 16/03/2020 @ 19:11:15.82: Initializing...

It's just a convenience feature and arguably not necessary for nerds logging everything ;) but perhaps it would be nice and hopefully not too difficult (?) to put some sort of separator in there.... even just a blank line or a row of some kind of spacer or something, so it's visually simple to tell one run from the next. I mean sure, they can just search for "Initializing..." or pay attention to the timestamps, but something like  this: 


    Mon 16/03/2020 @ 19:06:56.29: Cleaning temporary files...
    
    Mon 16/03/2020 @ 19:11:15.82: Initializing...

Would just be a nice touch.

**Comment by ScriptTiger** [comment]
> Wrapping %LOG% in quotes to "%LOG%" as such: (just these three instances, line 752, 759 and 760)

You are absolutely correct! I changed how I handle quotes halfway through the last update when I was extending the persistent logging functionality but forgot to finish the job. Just fixed it.

> I don't want to create even more work for you here but one thing I noticed was that the previous and current/next logs tend to run into each other a little in the log file

Just added this, too. Wasn't too much work, just one short line. I did, however, use "-----" as the separator instead of an empty line just to make it stand out a bit more, since there is already a lot of command output that logs empty lines and it might still be hard to find things.

Check out the last commit and let me know!

**Comment by ghost** [comment]
Works like a charm! 

I threw everything at it trying to bust it and the only thing that worked was a space in the log file name eg `/LOG:"This breaks.txt"` 
Long filenames without spaces work fine even without quotes eg `/LOG:thistotallyworkssomehow.txt`
.... so yay for that. 

Honestly, I think it's fine the way it is. I mean if you can think of a 5 second, zero-effort way to make it accept spaces in the log file name then great, but honestly... who would do that? XD I only noticed it because I'm an ex-test engy who was literally trying to break it. If an admin manages to do that by accident I'm not sure there's a patch for that kind of derp, LOL


Joking aside: Thanks again mate, great work as always.

**Comment by ScriptTiger** [comment]
> I threw everything at it trying to bust it and the only thing that worked was a space in the log file name

Fixed! Try it now. Admittedly I do have a bad habit of not catering to spaces and quotes, but it is a habit I do need to break.

> I mean if you can think of a 5 second, zero-effort way to make it accept spaces in the log file name then great

Lol, thanks for the consideration :P Yeah, this one was only one more line, as well, no biggie.

**Comment by ghost** [comment]
Hah you're a wizard mate. Works great. Names with spaces do require quotes but I think that's entirely expected behaviour. EG (and yes these are the actual file names I used, hahaha)

`/LOG:dont be stupid.txt `doesn't work
`/LOG:"dont be stupid.txt"` works

It's knock-off time mate I think you've earned a cold one.

**Comment by ScriptTiger** [comment]
Thanks for following this one to the end! @ohadschn, I know you're following, too, I think this might have cleaned up the last of your gripes, as well. So thanks, everybody, for the great feedback, as usual, and just pushed the new 1.39!

**Comment by ghost** [comment]
I'm sorry to be that annoying guy but I'm afraid that the update seems to be stuck in a loop. 

I just ran my 1.39 version scheduled task (with the edited line to say it's 1.38 from the testing above) and it tried to update and just got stuck in the loop. Created update.cmd but never renamed and ran it, just kept looping through and re-downloading the update.
Replaced it with the actual 1.38 and 1.39 versions from the repo and it did the same. 

For now I've just grabbed the master zip and removed and replaced everything and it's working fine but it might be worth mentioning for consideration during the next version update. You may recall that this thread began by my noticing that my script wasn't updating, so perhaps this is related. 

I had a very cursory glance at why it might be doing that and noticed that the rename of the file is dependent on the lock file existing and I notice that it's never created - as in, I never have a lock file.

Probably my stupid 'running it from a path with spaces' thing (Let's face it, still not a good idea on Windows even after all these years), rearing it's head again, lol. I'm tempted to just not do that any more but on the other hand I feel like maybe I'm useful as a guinea pig for others who might do it. Anyway, now that it's logging, I'll be more able to figure out what's happening....See you at 1.40 ;)

**Comment by ScriptTiger** [comment]
If there is something directly inhibiting the core functionality of the script, that's definitely a critical issue. Please keep me posted on your findings, I'll try to do some independent parallel testing of my own, as well.

**Comment by ghost** [comment]
I've found the issue - it's downloading the updated script, and it's running it, with the /U switch, but the /U switch is not detected, so it runs as normal, and then attempts to update itself over and over.

To demonstrate this we can just edit the update.cmd as such:

    rem Check if script is returning from being updated and finish update process
    echo "%1" &pause
    echo "%*" &pause
    if "%1"=="/U" (
    	cls
    	call :Echo "The updated script has been loaded"

The output will look like so:

> C:\Users\xcasxcursex\Documents\Hosts Update>Update.cmd /U
> Initializing...
> ""
> Press any key to continue . . .
> "/U"
> Press any key to continue . . .
> Terminate batch job (Y/N)? y


Uhm, wat?

So in truly haphazard form I tried a bunch of stuff and lo and behold:

> rem Check if script is returning from being updated and finish update process
> echo "%1" &pause
> echo "%~1" &pause
> echo "%0" &pause
> echo "%1" &pause
> echo "%2" &pause
> echo "%*" &pause
> echo exit here buddy &pause

outputs

> C:\Users\xcasxcursex\Documents\Hosts Update>Update.cmd /U
> Initializing...
> ""
> Press any key to continue . . .
> ""
> Press any key to continue . . .
> "/U"
> Press any key to continue . . .
> ""
> Press any key to continue . . .
> ""
> Press any key to continue . . .
> "/U"
> Press any key to continue . . .
> exit here buddy
> Press any key to continue . . .
> Terminate batch job (Y/N)?


So, yeh. It thinks the first switch is %0. But as far as I know, that's meant to be the actual command name so ....I'll stick to "uhm, wat?"

I feel like I could fix this and I'll probably do it poorly and take 10 times as long as you. Now that I've pinned down the cause of the endless update loop, I have a funny feeling you'll pull another 1-line-wizard trick and fix it, so I'll step aside and let the wizard do his magic, if you like.

**Comment by ScriptTiger** [comment]
Took a bit more than a line this time. This actually is a pretty serious issue, so let me know as soon as you get a chance to test this so I can push out the update ASAP.

**Comment by ghost** [comment]
Looks good. I didn't change the version number, just left it at 1.39. It saw the update, made it, ran it and replaced it and seems to be working fine.

Edit: I still never see a lock file but I'm not sure I should be in this situation? Anyway it seems fine.

**Comment by ghost** [comment]
Golden. Ran the pre 1.40 script (the 1.40 one that was still called 1.39 that I tested just before) and it updated and is now calling itself 1.40 and the VERSION file is reflecting it. Looks great.

Thanks man I'm sorry to cause all this work!

**Comment by ScriptTiger** [comment]
> Thanks man I'm sorry to cause all this work!

No, thank you. For me it's far from "extra work" to fix anything standing in the way of this script benefiting even a single person, for me that's just fulfilling the bare minimum in expectations. However, I have a feeling this particular issue definitely benefited a lot of other people, so definitely thank you immensely for following this through to the end.

I am scared to close it again in case I missed anything else, so I'll let you do the honors this time.

**Comment by ghost** [comment]
Hahahah I'm pretty sure you got it under control. I see from the commit why it thought %0 was the first arg, it was because it checked after the shift.

The only thing I'm not really sure about now is why I never see a lock file? When I was toying with it I actually managed to get it to create it one time (when I changed the %1 to %0), but it's entirely possible I was just breaking it.

Anyway, I feel like this particular issue is resolved (as well as the issue that I didn't even know existed which made me create this issue) .... and if the lock file is being created appropriately or not you can probably tell from your end so I'll close this and leave that in your obviously capable hands :)

**Comment by ScriptTiger** [comment]
> Edit: I still never see a lock file but I'm not sure I should be in this situation? Anyway it seems fine.

The lock file is created when you have previously set up a scheduled task and then subsequently activate that scheduled task from within an interactive script session. Since the scheduled task is run by the system account, the lock file just makes sure the system knows it's running as a slave to the interactive session and is thus not the only process using the script at this exact moment in time. Hence the name "lock," it's used to lock the script from being overwritten from an update or anything else the system might try if it starts to think it's the only one using the script. It's also used to pass the persistent logging information to the system process, as well. If the scheduled task is not configured for persistent logging but the current interactive script session is configured for persistent logging, the scheduled task will temporarily take the persistent logging configuration of the interactive script session.

**Comment by ghost** [comment]
> The lock file is created when you have previously set up a scheduled task and then subsequently activate that scheduled task from within an interactive script session. 

Aha So it is, and working as intended, thank you!
